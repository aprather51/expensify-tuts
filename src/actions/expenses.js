import uuid from 'uuid';
import database from '../firebase/firebase'

//ADD_EXPENSE *NEW* (For Firebase)
export const addExpense = (expense) =>({
    type: 'ADD_EXPENSE',
    expense
});

//ADD_EXPENSE *OLD*
// export const addExpense = ({ 
//     description = '', 
//     note = '', 
//     amount = 0, 
//     createAt = 0
// } = {}) =>({
//     type: 'ADD_EXPENSE',
//     expense:{
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createAt
//     }
// });



//** Setting up Expense data to dispatch based on *expenseData = {}*  **//
export const startAddExpense = (expenseData = {}) => {
    //Setting "return" only for redux-thunk... This function now get called interally by Redux and it get returned with argument called "dispatch"... 
    return (dispatch) => {
        //...And dispatch which gives an access inside items to write some data for firebase. 
        const { 
            description = '', 
            note = '', 
            amount = 0, 
            createAt = 0
        } = expenseData
        //set constant to select data from above that need to be pushed into firebase
        const expense = { description, note, amount, createAt }
    
        //firebase syntax to create/open array called 'expenses' and *push()* the variable called expense (see above "const expense = {...}) that contain items and values such as -- "description, note, amount and createAt" to firebase....The *then()* call back the success case for push to get called with "ref" together...
        return database.ref('expenses').push(expense).then((ref) => {
            //dispatching action called addExpense (Above) which will make changes when new data get updated.
            dispatch(addExpense({
                //dispatching a unique key with call name "ref"
                id: ref.key,
                //dispatching expense properties i.e. dispatch, note, amount, and createAt 
                ...expense
                //at end it pushes id based on .then(ref) and expense properities together.
            }));
        });
    };
};//End of Export
   

//REMOVE_EXPENSE
export const removeExpense = ({id} = {}) =>({
    type: 'REMOVE_EXPENSE',
    id
});

//
export const startRemoveExpense = ({id} = {}) => { 
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        })
    };
};

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
    type: 'EDIT_EXPENSE',
    id,
    update
});

//SET_EXPENSE
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//FETCH THE DATA AT START
export const startSetExpenses =  () => {
    /*Return with dispatch variable to get darta */
    return (dispatch) => {
        /* Return with promise string. Collect data reference called expenses from firebase with value 
        at once. When data is successfully obtained, it proceed to then() call with snapshot variable.
        The snapshot variable snaps at object structure.*/
        return database.ref('expenses').once('value').then((snapshot) => {
            
            /* setting up expenses with nothing */
            const expenses = [];

            /* Collect object structure and show list of items in array, Since snapshot isn't working
            best with redux. childSnapshot is get called as variable.*/
            snapshot.forEach((childSnapshot) => {
                /* To push() childSnapshot propties such as key and values to firebase*/
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            /* dispatching setExpense action with expense variable*/
            dispatch(setExpenses(expenses))
        });
       
    };
};