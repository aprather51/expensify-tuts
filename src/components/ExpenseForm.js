import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


// const now = moment();
// console.log(now.format('dddd, MMMM Do YYYY'))
// console.log(now.format('h:mm a'))

export default class ExpenseForm extends Component {

    //* Choice One
    // constructor(props) {
    //     super(props);

    //     this.state ={ 
    //         description: props.expense ? props.expense.description : '',
    //         note: props.expense ? props.expense.note : '',
    //         amount: props.expense ? (props.expense.amount / 100).toString() : '',
    //         createAt: props.expense ? moment(props.expense.createAt) : moment(),
    //         calendarFocused: false,
    //         error: ''
    //     };
    // }

    //* Choice 2 - (Prefered)
    state = { 
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createAt: this.props.expense ? moment(this.props.expense.createAt) : moment(),
        calendarFocused: false,
        error: ''
    };
    
    

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    
    //This should set correct amount value i.e. 23.21 not 23.212
    onAmountChange = (e) => {
        const amount = e.target.value;
        // (!amount || amount.match(....)) -- help to delete the value inside input
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }   
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDateChange = (createAt) => {
        // if conditional (no else or else if) -- help to prevent deleting date value inside input
        if(createAt){ 
            this.setState(() => ({ createAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (e) => {
         e.preventDefault();

        // if(!this.state.description || !this.state.amount) {
        //     this.setState(() => ({
        //         error: 'Please Provide description and Amount'
        //     }))
        //     console.log(this.setState.error);
        // } else {
        //     console.log('Submitted!');
        //     this.props.onSubmit({
        //         description: this.state.description,
        //         note: this.state.note,
        //         amount: parseFloat(this.state.amount, 10) * 100,
        //         //! base 10 )* 100 - converted into cents.
        //         createAt: this.state.createAt.valueOf() 
        //         //! .valueOf() moment unix timestamp.
        //     });
        // }

        if(!this.state.description && !this.state.amount ) {
            this.setState(() => ({
                error: 'Please include your description and amount'
            }));
        } else if (!this.state.description) {
            this.setState(() => ({
                error: 'Please include your description'
            }));
        } else if (!this.state.amount) {
            this.setState(() => ({
                error: 'Please include your amount'
            }));
        } else {
            this.setState(() => ({ error: '' }));
                
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                //! base 10 )* 100 - converted into cents.
                createAt: this.state.createAt.valueOf() 
                //! .valueOf() moment unix timestamp.
            });
        }; 
    }


    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.error && <p>{this.state.error}</p>}
                    <input 
                        type="text"
                        placeholder="description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />  
                    <input 
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="(Optional) Note"
                        cols="30" 
                        rows="10"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

// export default ExpenseForm;
