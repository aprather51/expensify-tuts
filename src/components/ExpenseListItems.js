import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItems = ({id, description, amount, createAt}) => (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description} </h3>
      </Link> 
      <p>
        {numeral(amount / 100).format('$0,0.00')} 
        - 
        {moment(createAt).format('dddd, MMMM D YYYY')}</p>
      <hr/>
    </div>
    
);

export default ExpenseListItems;