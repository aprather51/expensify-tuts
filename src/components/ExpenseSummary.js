import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expense';
import selectExpensesTotal from '../selectors/expenseTotal';

export const ExpenseSummary = ({expenseCount, expenseTotal}) =>{
    //Terms - no s at end of expense if only one item. Otherwise add s
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    //setting up grand total
    const grandTotal = numeral(expenseTotal / 100).format('$0,0.00');
    return (
     <div>
         <h3> Viewing {expenseCount} {expenseWord} totaling {grandTotal}</h3>
     </div>
    )};   

    const mapStateToProps = (state) => {
        //Set an constant -- to state visible of total amount expenses prior to each items, This also will state changes when amount is filtered out by date, amount or whatever is labeled inside input. 
       const visibleExpenses = selectExpenses(state.expenses, state.filters);
       //return with Numbers of items and total amount. 
       return{
           expenseCount: visibleExpenses.length,
           expenseTotal: selectExpensesTotal(visibleExpenses)
       }
    };

    export default connect(mapStateToProps)(ExpenseSummary);
