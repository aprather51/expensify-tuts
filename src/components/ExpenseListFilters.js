import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends Component{
    state = {
        calendarFocused: null //as string instead of t/f
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = ( calendarFocused) => {
        this.setState(() => ({ calendarFocused}));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        if(e.target.value === 'date'){  
        //If value on date is matched, then it will filter out the date.
            this.props.sortByDate();
        }else if(e.target.value === 'amount'){
         //else if the value on amount is matched, then it will filter out the amount.
            this.props.sortByAmount();
        }
    }
    
    render(){
        return(
            <div>
            <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
            <select 
               value={this.props.filters.sortBy} 
               onChange={this.onSortChange}
            >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            {/* Moment() setup */}
            <DateRangePicker 
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={ this.onFocusChange}
                showClearDates={true}  //hit X to clear the date
                numberOfMonths={1} //show only one month
                isOutsideRange={() => false} //show date in past without greyed out.  
            />
        </div>
        );
    }
}


const mapStateToProps = (state) =>({ filters: state.filters });

const mapDispatchToProps = (dispatch) =>({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStarDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);