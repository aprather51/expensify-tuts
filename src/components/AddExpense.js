import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends Component{
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  }
  render(){
    return(
      <div>
      <h1>Add Expense</h1>
      <ExpenseForm 
          onSubmit={this.onSubmit}
      />
    </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});
export default connect(undefined, mapDispatchToProps )(AddExpensePage);

// // const AddExpensePage = (props) => (
// //     <div>
// //       <h1>Add Expense</h1>
// //       <ExpenseForm 
// //          onSubmit={(expense) =>{
// //            //props.dispatch(addExpense(expense)); //from ExpenseForm.js (onSubmit function) to react-redux

// //           props.onSubmit(expense)

// //           props.history.push('/'); //redirect to root page after submit.
// //         }}
// //     />
// //   </div>
// //);



