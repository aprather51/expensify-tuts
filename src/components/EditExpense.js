import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense  } from '../actions/expenses';

export class EditExpensePage extends Component{
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render(){
        return(
            <div>
            <h1>Edit Expense</h1>
            {/* Edit Expense Component  - The id => {props.match.params.id} */}
            <ExpenseForm 
                expense={this.props.expense}
                onSubmit={this.onSubmit}
            />
            <button onClick={this.onRemove}>Remove</button>
        </div>
        )
    }
}
// // const EditExpensePage = (props) =>{ 
// //     console.log(props)
// //     return(
// //         <div>
// //             <h1>Edit Expense</h1>
// //             {/* Edit Expense Component  - The id => {props.match.params.id} */}
// //             <ExpenseForm 
// //                 expense={props.expense}
// //                 onSubmit={(expense) => {
// //                     console.log('updated!', expense)
// //                     props.dispatch(editExpense(props.expense.id, expense))  
// //                     //Id matched item (props.expense,id) then make change (expense)
// //                     props.history.push('/')
// //                     //Redirect
// //                 }} 
// //             />
// //             <button onClick={() => {
// //                 props.dispatch(removeExpense({ id: props.expense.id }))
// //                 props.history.push('/')
// //             }}>Remove</button>
// //         </div>
// //     );
// // };


// mapStateToProps allows us define what values from the Redux store the component should have access to. We can use it to give the component access to the entire store or just a specific subset of it's data.
const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),

});

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(undefined, mapStateToProps, mapDispatchToProps)(EditExpensePage);

