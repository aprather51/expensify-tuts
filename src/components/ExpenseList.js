import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItems from './ExpenseListItems';
import selectExpense from '../selectors/expense';


export const ExpenseList = (props) => (
    // <div>
    //     <h1>Expense List</h1>
    //     {props.expenses.map((expense) => {
    //         return <ExpenseListItems key={expense.id} {...expense}/>
    //     })}
    // </div>
    <div>
        {
            props.expenses.length === 0 ? (
                <h1>Expense List</h1>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItems key={expense.id} {...expense}/>
                })
            )
        }

    
    </div>  
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseList);

