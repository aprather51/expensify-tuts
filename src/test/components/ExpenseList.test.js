import React from 'react'
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixures/expenses';

test('Render Expense List with expenses',() =>{
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should Render Expense List with empty message expenses',() =>{
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});

