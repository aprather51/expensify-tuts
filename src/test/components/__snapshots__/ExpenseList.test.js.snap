// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Expense List with expenses 1`] = `
<div>
  <ExpenseListItems
    amount={195}
    createAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItems
    amount={109500}
    createAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItems
    amount={4500}
    createAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`Should Render Expense List with empty message expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
</div>
`;
